File: app/src/main/java/ru/SnowVolf/convertx/regexdragon/Flags.java
package ru.SnowVolf.convertx.regexdragon;

import java.util.ArrayList;

/**
 * Created by Snow Volf on 26.02.2017, 11:29
 */

public class Flags {
    private ArrayList<Boolean> mData = new ArrayList<>();
    private CharSequence[] mOptions = new CharSequence[]{"Case insensitive [i]", "Multiline [m]", "Comments [x]", "Dotall [s]", "Literal [l]", "Unicode Case [u]", "Unix Lines [d]"};

    public Flags(){
        for (int i = 0; i < mOptions.length; i++){
            mData.add(i, false);
        }
    }

    public CharSequence[] getOptions(){
        return mOptions;
    }

    public void add(int id){
        mData.set(id, true);
    }

    public void remove(int id){
        mData.set(id, false);
    }

    public boolean[] getSelectedBooleans(){
        boolean[] data = new boolean[mOptions.length];
        for (int i = 0; i < mData.size(); i++){
            data[i] = mData.get(i);
        }
        return data;
    }
    /**
     * Чтобы не забыть
     * CANON_EQ (no effect on Android) == 128
     * CASE_INSENSITIVE == 2
     * COMMENTS == 4
     * DOTALL == 32
     * LITERAL == 16
     * MULTILINE == 8
     * UNICODE_CASE == 64
     * UNICODE_CHARACTER_CLASS (no effect on Android) == 256
     * UNIX_LINES == 1
     *
     * все константы можно посмотреть на developer.android.com
     */

    public int getFlags(){
        int data = 0;
        if (mData.get(0)){
            data = 2;
        }
        if (mData.get(1)){
            return data | 8;
        }
        if (mData.get(2)){
            return data | 4;
        }
        if (mData.get(3)){
            return data | 32;
        }
        if (mData.get(4)){
            return data | 16;
        }
        if (mData.get(5)){
            return data | 64;
        }
        if (mData.get(6)){
            return data | 1;
        }
        return data;
    }

    public String getFlagString(){
        String data = "/";
        if (mData.get(0)){
            data = data + "i";
        }
        if (mData.get(1)){
            data = data + "m";
        }
        if (mData.get(2)){
            data = data + "x";
        }
        if (mData.get(3)){
            data = data + "s";
        }
        if (mData.get(4)){
            data = data + "l";
        }
        if (mData.get(5)){
            data = data + "u";
        }
        if (mData.get(6)){
            data = data + "d";
        }
        if (data.equals("/")){
            return "Флаги";
        }
        return data;
    }
}

File: app/src/main/java/ru/svolf/convertx/regexdragon/Flags.kt
package ru.svolf.convertx.regexdragon

import java.util.*

/**
 * Created by Snow Volf on 26.02.2017, 11:29
 */

internal class Flags {
    private val mData = ArrayList<Boolean>()
    val options = arrayOf("Case insensitive [i]", "Multiline [m]", "Comments [x]", "Dot all [s]", "Literal [-]", "Unicode Case [u]", "Unix Lines [d]")

    init {
        for (i in options.indices) {
            mData.add(i, false)
        }
    }

    fun add(id: Int) {
        mData[id] = true
    }

    fun remove(id: Int) {
        mData[id] = false
    }

    val selectedBooleans: BooleanArray
        get() {
            val data = BooleanArray(options.size)
            for (i in mData.indices) {
                data[i] = mData[i]
            }
            return data
        }
    /**
     * Чтобы не забыть
     * CANON_EQ (no effect on Android) == 128
     * CASE_INSENSITIVE == 2
     * COMMENTS == 4
     * DOTALL == 32
     * LITERAL == 16
     * MULTILINE == 8
     * UNICODE_CASE == 64
     * UNICODE_CHARACTER_CLASS (no effect on Android) == 256
     * UNIX_LINES == 1

     * все константы можно посмотреть на developer.android.com
     */

    val flags: Int
        get() {
            var data = 0
            if (mData[0]) {
                data = 2
            }
            if (mData[1]) {
                return data or 8
            }
            if (mData[2]) {
                return data or 4
            }
            if (mData[3]) {
                return data or 32
            }
            if (mData[4]) {
                return data or 16
            }
            if (mData[5]) {
                return data or 64
            }
            if (mData[6]) {
                return data or 1
            }
            return data
        }

    val flagString: String
        get() {
            var data = "/"
            if (mData[0]) {
                data += "i"
            }
            if (mData[1]) {
                data += "m"
            }
            if (mData[2]) {
                data += "x"
            }
            if (mData[3]) {
                data += "s"
            }
            if (mData[4]) {
                data += "-"
            }
            if (mData[5]) {
                data += "u"
            }
            if (mData[6]) {
                data += "d"
            }
            if (data == "/") {
                return "Флаги"
            }
            return data
        }

}
